src/System.IO.FileSystem.AccessControl/src/System/IO/FileSystemAclExtensions.cs

```diff
@@ -10,6 +10,20 @@ namespace System.IO
 {
     public static class FileSystemAclExtensions
     {
+        /// <summary>Creates a new directory, ensuring it is created with the specified directory security. If the directory already exists, nothing is done.</summary>
+        /// <param name="directoryInfo">The object describing a directory that does not exist in disk yet.</param>
+        /// <param name="directorySecurity">An object that determines the access control and audit security for the directory.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="directorySecurity" /> is <see langword="null" />.</exception>
+        /// <exception cref="DirectoryNotFoundException">Could not find a part of the path.</exception>
+        /// <exception cref="UnauthorizedAccessException">Access to the path is denied.</exception>
+        public static void Create(this DirectoryInfo directoryInfo, DirectorySecurity directorySecurity)
+        {
+            if (directorySecurity == null)
+                throw new ArgumentNullException(nameof(directorySecurity));
+
+            FileSystem.CreateDirectory(directoryInfo.FullName, directorySecurity.GetSecurityDescriptorBinaryForm());
```

Comment:
This will null ref if directoryInfo is null. Don't we usually throw ArgumentNullException for such cases, even for extension methods?

Reply:
I didn't know extension methods could be called from a null argument that was passed as `this`. None of the other extension methods in this class do a null check. What would be the correct thing to do then?

Reply:
I've been reading about this. I will definitely add the check to throw ArgumentNullException if null.
@stephentoub @JeremyKuhne  Should I add the same check to the existing extension methods?

Reply:
Yes, but that can happen in 5.0

Reply:
I opened an issue to track this: https://github.com/dotnet/corefx/issues/41907

Reply:
Thanks. I don't actually have a strong preference on these; we should just be consistebt.

@terrajobst, @bartonjs, what do our design guidelines say?

Reply:
> what do our design guidelines say [for `null` in extension methods]?

Nothing written down.  The problem is (syntactically speaking) from an extension invocation it should be NullReferenceException and from static invocation it should be ArgumentNullException.  Given that we don't know which call style it was, the proposed guidelines I have are

* AVOID defining extension methods that accept a `null` value for a `this` parameter.
* DO NOT allow the `this` parameter being `null` produce a NullReferenceException.

Though, having sat on that for a while, I can now edit it to a much cleaner

* DO throw `ArgumentNullException` when the `this` parameter is `null` in an extension method.

[(null).Something() succeeding confuses people, don't nullref because of static calling conventions, et cetera]

Reply:
I don't have a strong preference either, but my assumption is that extension methods are almost always called as "extension invocation" which would suggest NRE would be less confusing. But, my impression is that we are generally using ANE in our own code today.

Reply:
Counter argument -- if you are using "static invocation" you might be glad to have the parameter name. Anyway - I say go with whatever @bartonjs writes down 😸

Reply:
Thanks everyone.